File: qna/src/main/java/com/example/rag/qna/domain/cache/CategoryQuestionTitleCache.java
package com.example.rag.qna.domain.cache;

import core.rag.QuestionTitle;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

@Component
public class CategoryQuestionTitleCache {
    private final Map<String, QuestionTitleCache> caches;

    CategoryQuestionTitleCache() {
        caches = new HashMap<>();
    }

    public QuestionTitle[] getRecentQuestionTitles(String category, int startNum, int endNum) {
        caches.computeIfAbsent(category, k -> new QuestionTitleCache(1000));
        return caches.get(category).getRecentQuestionTitles(startNum, endNum);
    }

    public void putRecentQuestionTitle(String category, QuestionTitle questionTitle) {
        caches.computeIfAbsent(category, k -> new QuestionTitleCache(1000));
        caches.get(category).putRecentQuestionTitle(questionTitle);
    }
}

File: qna/src/main/kotlin/com/example/rag/qna/domain/cache/CategoryQuestionTitleCache.kt
package com.example.rag.qna.domain.cache

import core.rag.QuestionTitle
import org.springframework.stereotype.Component

@Component
class CategoryQuestionTitleCache {
    private val caches: MutableMap<String, QuestionTitleCache> = HashMap()

    fun getRecentQuestionTitles(category: String, startNum: Int, endNum: Int): Array<QuestionTitle?>? {
        caches.computeIfAbsent(category) { k: String? -> QuestionTitleCache() }
        return caches[category]?.getRecentQuestionTitles(startNum, endNum)
    }

    fun putRecentQuestionTitle(category: String, questionTitle: QuestionTitle?) {
        caches.computeIfAbsent(category) { k: String? -> QuestionTitleCache() }
        caches[category]!!.putRecentQuestionTitle(questionTitle)
    }
}